### ISSUE TITLE
Assembler: Add Label.Rip

### ISSUE THREAD

**Issue by Symbai** [0xd4d/iced#63]
Adding a property to return the location of the label. This is extremely useful when it is used elsewhere again. The assembler must know it, so it only needs to store/expose this information.

```
var label1 = c.CreateLabel();
c.push(r15);
c.add(rax, r15);
c.jne(label1);
c.inc(rax);
c.Label(label1);
c.pop(r15);
c.ret();
c.Encode();
```
Then later:
```
var c = Assembler.Create(...)
var rip = label1.Rip // == 0xB
c.jmp(label1)
c.Encode();
```

Another situation:
```
var stream = new MemoryStream();
var c = Assembler.Create(64, new StreamCodeWriter(stream));
var label1 = c.CreateLabel();
c.mov(rax, __[label1]);
c.jmp(0x40000)
c.Label(label1);
c.dq(4d);
c.Encode();
WriteProcessMemory(...., ...., stream.ToArray(), ....)
```
Then later:
```
var rip = label1.Rip
WriteProcessMemory(..., rip, BitConverter.GetBytes(2d), ...)
```

**Comment by xoofx** [comment]
It is not really possible to do that because `Label` as `Instruction` is a struct, so it's not a reference.

You can recover the location of a label like that though (as you would do to recover the location of any instructions)

```
		// Emit label1:
		c.Label(label1);
		var label1Index = c.Instructions.Count;
		c.pop(r15);
		c.ret();

		var result = c.Encode(BlockEncoderOptions.ReturnNewInstructionOffsets);

		var label1Offset = result.NewInstructionOffsets[label1Index];
```

- Get the index of the instruction on the label
- Encode with `BlockEncoderOptions.ReturnNewInstructionOffsets`
- Recover the offsets via `encodingResult.NewInstructionOffsets`

**Comment by Symbai** [comment]
This seems to be a lot of afford when assembling regularly, consider we assemble something with 4-20 labels. Wouldn't it be possible that what you are doing in your example can be done by the assembler? It seems to only need to remember the label positions and once it processed a c.Label(...) line it can store the position internally by calling c.Instructions.Count, like you did in your example.

Maybe there can be an Assembler.Create() overload with assembler settings to optionally turn this on?

**Comment by xoofx** [comment]
> This seems to be a lot of afford when assembling regularly

I agree. I think I should be able to come up with a change in the API like:

```
// Emit label1:
c.Label(ref label1);
c.pop(r15);
c.ret();

var result = c.Encode(0, BlockEncoderOptions.ReturnNewInstructionOffsets);
var label1RIP = result.GetLabelRIP(label1);
```

Result will be something like `AssemblerResult` instead, 0 is the BaseRIP passed by parameter to the encoder. You will need to pass the label by ref to `Label` and it will store an `InstructionIndex` in it. The InstructionIndex can then be retrieved by `GetLabelRIP`... This is the least intrusive, concise and most efficient way I can think of.

**Comment by 0xd4d** [comment]
Branches can get optimized and if the branch is too far away for a normal branch, it will get rewritten to 2+ instructions instead, eg. `je too_far` in 64-bit mode becomes `jne short skip; jmp qword ptr [mem]`. In that case, the offset of that instruction isn't returned since it's not part of the output. Instead `uint.MaxValue` is returned in that array.

We could add a new option to always get instruction offsets even if the instruction got re-written to 2+ other instructions.

**Comment by 0xd4d** [comment]
> I will come later today with a PR if @0xd4d agree on this proposal.

Yes, send the PR :)

**Comment by xoofx** [comment]
> Branches can get optimized and if the branch is too far away for a normal branch, it will get rewritten to 2+ instructions instead, eg. je too_far in 64-bit mode becomes jne short skip; jmp qword ptr [mem]. In that case, the offset of that instruction isn't returned since it's not part of the output. Instead uint.MaxValue is returned in that array.

Btw, I was very surprised by this behavior. Is it something that other assemblers are also doing?

**Comment by 0xd4d** [comment]
> Btw, I was very surprised by this behavior. Is it something that other assemblers are also doing?

Do you mean re-writing a branch that is too far away? I don't know.

The alternative would be to fail, which I don't like since it's possible to fix it by negating the branch and adding a `jmp [mem]` instruction.

**Comment by xoofx** [comment]
Fixed by PR #66
